### PR TITLE
[MAINT] Refactor/test utils

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Relates to #3570

Changes proposed in this pull request:

- refactor complex tests in nilearn/_utils/tests

- [x] more refactoring to do once #3905 is in

### CODE REVIEW

**Comment by Remi-Gau** [comment]
more refactoring to do once #3905 is in

File: nilearn/_utils/tests/test_numpy_conversions.py [hunk 1]
```diff
@@ -126,25 +125,30 @@ def test_as_ndarray():
...
-        else:
-            assert arr2.dtype == out_dtype, str(case)
-
-        result_order = out_order if out_order is not None else in_order
-        if result_order == "F":
-            assert arr2.flags["F_CONTIGUOUS"], str(case)
-        else:
-            assert arr2.flags["C_CONTIGUOUS"], str(case)
-
+    # for case in test_cases:
```
Comment (bthirion) [inline nilearn/_utils/tests/test_numpy_conversions.py:134]: to be cleaned

**Review by bthirion — COMMENTED** [review]
LGTM so far.

File: nilearn/_utils/tests/test_numpy_conversions.py (hunk 1 above)
Comment (Remi-Gau) [reply nilearn/_utils/tests/test_numpy_conversions.py]: tried to use more transparent variable names and add comment to explain what is happening

File: nilearn/_utils/tests/test_niimg_conversions.py [hunk 2]
```diff
@@ -273,72 +273,68 @@ def test_check_niimg_wildcards():
...
-    # No globbing behavior
-    with testing.write_tmp_imgs(img_4d, create_files=True) as filename:
-        assert_array_equal(
-            get_data(_utils.check_niimg(filename, wildcards=False)),
-            get_data(img_4d),
-        )
 
-    #######
+def test_check_niimg_wildcards_one_file_name(img_3d_zeros_eye):
+    tmp_dir = tempfile.tempdir + os.sep
```
Comment (ymzayek) [inline nilearn/_utils/tests/test_niimg_conversions.py:293]: should we replace with tmp_path fixture?

File: nilearn/_utils/tests/test_niimg_conversions.py [hunk 3]
```diff
@@ -383,17 +380,20 @@ def test_iter_check_niimgs(tmp_path):
...
+def test_iter_check_niimgs(tmp_path, img_4d_zeros_eye):
+    img_2_4d = [[img_4d_zeros_eye, img_4d_zeros_eye]]
+
     # Create a test file
     fd, filename = tempfile.mkstemp(
         prefix="nilearn_test", suffix=".nii", dir=str(tmp_path)
     )
     os.close(fd)
-    img_4d.to_filename(filename)
+    img_4d_zeros_eye.to_filename(filename)
```
Comment (ymzayek) [inline nilearn/_utils/tests/test_niimg_conversions.py:392]: Is it possible to use the context manager ` testing.write_tmp_imgs` here like in some tests above?

**Review by ymzayek — COMMENTED** [review]
LGTM, just a couple of comments

File: nilearn/_utils/tests/test_niimg_conversions.py (hunk 3 above)
Comment (Remi-Gau) [reply nilearn/_utils/tests/test_niimg_conversions.py]: done

File: nilearn/_utils/tests/test_niimg_conversions.py (hunk 2 above)
Comment (Remi-Gau) [reply nilearn/_utils/tests/test_niimg_conversions.py:293]: tried the most obvious ways and they did not work

thinking that maybe we could improve / fixturize 'testing.write_tmp_imgs', but that would be for another PR

**Comment by Remi-Gau** [comment]
OK it seems that https://github.com/nilearn/nilearn/pull/3910/commits/244f0f38913b613a8b52f3e926d354ff37d0dfa7 does not play well with testing on windows

File: nilearn/_utils/tests/test_niimg_conversions.py (hunk 3 above)
Comment (Remi-Gau) [reply nilearn/_utils/tests/test_niimg_conversions.py]: OK it seems that this does not play well with testing on windows

Comment (Remi-Gau) [reply nilearn/_utils/tests/test_niimg_conversions.py]: will revert for now

File: nilearn/_utils/tests/test_niimg_conversions.py (hunk 2 above)
Comment (ymzayek) [reply nilearn/_utils/tests/test_niimg_conversions.py:293]: Ok agreed

**Review by ymzayek — APPROVED** [review]
Thx @Remi-Gau

**Pull request merged**